dbms/src/Functions/FunctionsLogical.h

```diff
@@ -587,13 +654,23 @@ class FunctionAnyArityLogical : public IFunction
         else
         {
             const auto & col_data = nullable_uint8_columns[0]->getData();
-            // according to https://github.com/pingcap/tiflash/issues/5849
-            // need to cast the UInt8 column to bool explicitly
-            for (size_t i = 0; i < rows; ++i)
+            if constexpr (null_as_false)
             {
-                vec_res[i] = static_cast<bool>(col_data[i]);
+                for (size_t i = 0; i < rows; ++i)
+                {
+                    vec_res[i] = !(*null_maps[0])[i] && static_cast<bool>(col_data[i]);
+                }
+            }
+            else
+            {
+                // according to https://github.com/pingcap/tiflash/issues/5849
+                // need to cast the UInt8 column to bool explicitly
```

Comment:
These comments can be moved before `if constexpr (null_as_false)`

Reply:
Done